### PR TITLE
[MNG-7547] Simpler G level metadata generation

### PR DESCRIPTION
**Pull request by cstamas**
Just inspect deployed artifacts, try to find a plugin among them and reuse PluginDescriptor embedded in it.

---

https://issues.apache.org/jira/browse/MNG-7547

### CODE REVIEW

File: maven-resolver-provider/pom.xml [hunk 1]
```diff
@@ -34,6 +34,10 @@ under the License.
   <description>Extensions to Maven Resolver for utilizing Maven POM and repository metadata.</description>
 
   <dependencies>
+    <dependency>
+      <groupId>org.apache.maven</groupId>
+      <artifactId>maven-plugin-api</artifactId>
+    </dependency>
```
Comment (cstamas) [inline maven-resolver-provider/pom.xml:40]: Needed to grab PluginDescriptor model, that needs plexus... :cry:

File: maven-resolver-provider/pom.xml [hunk 2]
```diff
@@ -103,6 +107,13 @@ under the License.
       <artifactId>failureaccess</artifactId>
       <optional>true</optional>
     </dependency>
+
+    <!-- Is must due PluginDescriptor as it uses Plexus stuff -->
+    <dependency>
+      <groupId>org.eclipse.sisu</groupId>
+      <artifactId>org.eclipse.sisu.plexus</artifactId>
```
Comment (cstamas) [inline maven-resolver-provider/pom.xml:114]: Needed for model and as builder throws PlexusConfEx :sob: 

Maybe just slurp in the plugin.xml as DOM?

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java [hunk 3]
```diff
@@ -124,7 +128,46 @@ public Collection<? extends Metadata> finish( Collection<? extends Artifact> art
...
+                        return new PluginInfo(
+                                pluginDescriptor.getGroupId(),
+                                pluginDescriptor.getArtifactId(),
+                                pluginDescriptor.getGoalPrefix(),
+                                pluginDescriptor.getName() );
+                    }
+                }
+                catch ( Exception e )
+                {
+                    // ignore it for now
```
Comment (cstamas) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:167]: This is too broad, but for example maven's ArtifactDeployerTest and ArtifactInstallerTest (maven-compat) uses invalid JAR for tests.... :sob:

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java [hunk 4]
```diff
@@ -124,7 +128,46 @@ public Collection<? extends Metadata> finish( Collection<? extends Artifact> art
...
+                    ZipEntry pluginDescriptorEntry = artifactJar.getEntry( PLUGIN_DESCRIPTOR_LOCATION );
+
+                    if ( pluginDescriptorEntry != null )
+                    {
+                        InputStream is = artifactJar.getInputStream( pluginDescriptorEntry );
+
+                        Reader reader = ReaderFactory.newXmlReader( is );
+
+                        PluginDescriptor pluginDescriptor =
+                                pluginDescriptorBuilder.build( reader, artifactPath.toAbsolutePath().toString() );
```
Comment (cstamas) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:156]: Maybe just load up DOM and drop plugin-api/plexus as deps?

**Comment by gnodet** [comment]
The branch http://github.com/gnodet/maven/tree/plugin-metadata fixes the conflict and also remove the metadata from the v4 api.

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadata.java [hunk 5]
```diff
@@ -27,15 +27,32 @@
...
 /**
- * Plugin G level metadata.
+ * Maven G level metadata.
  */
 final class PluginsMetadata
     extends MavenMetadata
 {
+    static final class PluginInfo
+    {
+        final String groupId;
```
Comment (michael-o) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadata.java:39]: Why is this field not private? Maybe leave a comment for this?

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java [hunk 6]
```diff
@@ -35,42 +43,35 @@
...
+ * Plugin metadata contains G level list of "prefix" to A mapping for plugins present under this G.
  */
 class PluginsMetadataGenerator
-    implements MetadataGenerator
+        implements MetadataGenerator
 {
-    private final PluginsMetadataInfoProvider pluginsMetadataInfoProvider;
-
-    private final Map<Object, PluginsMetadata> plugins;
+    private static final String PLUGIN_DESCRIPTOR_LOCATION = "META-INF/maven/plugin.xml";
```
Comment (michael-o) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:54]: `/META-INF` or `META-INF`?

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java [hunk 7]
```diff
@@ -124,7 +126,50 @@ public Collection<? extends Metadata> finish( Collection<? extends Artifact> art
...
+    private PluginInfo extractPluginInfo( Artifact artifact )
+    {
+        // sanity: jar, no classifier and file exists
+        if ( artifact != null
+                && "jar".equals( artifact.getExtension() )
+                && "".equals( artifact.getClassifier() )
+                && artifact.getFile() != null )
+        {
+            Path artifactPath = artifact.getFile().toPath();
+            if ( Files.isRegularFile( artifactPath ) )
```
Comment (michael-o) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:141]: What if the plugin is the reactor and the jar hasn't been built? Can this happen?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java (hunk 6 above)
Comment (cstamas) [reply maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:54]: Why `/META-INF`?

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java (hunk 7 above)
Comment (cstamas) [reply maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:141]: PluginsMetadataGenerator <-isInvokedBy- PluginsMetadataGeneratorFactory <-isInvokedBy- Deployer... hence, your question is "what if a deploy is ongoing for a project that is not built" that is a contradiction (or just bad API usage of resolver?)

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadata.java (hunk 5 above)
Comment (cstamas) [reply maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadata.java:39]: the whole class is pkg private, is shared by two classes from this package

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java (hunk 6 above)
Comment (michael-o) [reply maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:54]: All good, it is consistent with the rest of our code.

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java (hunk 7 above)
Comment (michael-o) [reply maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java:141]: Thanks, that is what I wanted to hear. We won't get so far w/o building the JAR.

**Comment by michael-o** [comment]
@cstamas Can you please rebase? I want to test this as well.

**Comment by cstamas** [comment]
@michael-o rebased

**Comment by michael-o** [comment]
Going through...

**Review by michael-o — APPROVED** [review]

**Pull request merged**